subsys/net/lib/dns/resolve.c

```diff
@@ -1271,10 +1302,14 @@ static int dns_resolve_close_locked(struct dns_resolve_context *ctx)
 {
 	int i;
 
-	if (!ctx->is_used) {
+	if (ctx->state != DNS_RESOLVE_CONTEXT_ACTIVE) {
 		return -ENOENT;
 	}
 
+	ctx->state = DNS_RESOLVE_CONTEXT_DEACTIVATING;
+
+	k_mutex_unlock(&ctx->lock);
```

Comment:
Why do we unlock, after all for loop is still accessing the ctx fields.

Reply:
Good question. Added following comment:
```
	/*
	 * ctx->net_ctx is never used in "deactivating" state. Additionally
	 * following code is guaranteed to be executed only by one thread at a
	 * time, due to required "active" -> "deactivating" state change. This
	 * means that it is safe to put net_ctx with mutex released.
	 *
	 * Released mutex will prevent lower networking layers from deadlock
	 * when calling cb_recv() (which acquires ctx->lock) just before closing
	 * network context.
	 */
	k_mutex_unlock(&ctx->lock);
```

Reply:
ok